code/game/objects/structures/tank_dispenser.dm

```diff
@@ -88,13 +88,13 @@
 	switch(action)
 		if("plasma")
 			var/obj/item/weapon/tank/internals/plasma/tank = locate() in src
-			if(tank)
+			if(tank && Adjacent(usr))
 				usr.put_in_hands(tank)
 				plasmatanks--
 			. = TRUE
 		if("oxygen")
 			var/obj/item/weapon/tank/internals/oxygen/tank = locate() in src
-			if(tank)
+			if(tank && Adjacent(usr))
 				usr.put_in_hands(tank)
 				oxygentanks--

```

Comment:
```
else if(tank)
(indent)tank.move(src.get_turf())

```
Indented to the same level as the if()

You could also use `src.loc` if you always expect these be on a floor or you want it to end up in the same spot as the dispenser(like if a mech picked it up inside the mech)